jetty-core/jetty-siwe/src/main/java/org/eclipse/jetty/security/siwe/SignInWithEthereumParser.java

```diff
@@ -0,0 +1,65 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.security.siwe;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class SignInWithEthereumParser
+{
+    private static final String SCHEME_PATTERN = "[a-zA-Z][a-zA-Z0-9+\\-.]*";
+    private static final String DOMAIN_PATTERN = "(?:[a-zA-Z0-9\\-._~%]+@)?[a-zA-Z0-9\\-._~%]+(?:\\:[0-9]+)?";
+    private static final String ADDRESS_PATTERN = "0x[0-9a-fA-F]{40}";
+    private static final String STATEMENT_PATTERN = "[^\\n]*";
+    private static final String URI_PATTERN = "[^\\n]+";
+    private static final String VERSION_PATTERN = "[0-9]+";
+    private static final String CHAIN_ID_PATTERN = "[0-9]+";
+    private static final String NONCE_PATTERN = "[a-zA-Z0-9]{8}";
+    private static final String DATE_TIME_PATTERN = "\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(?:\\.\\d+)?(?:Z|[+-]\\d{2}:\\d{2})?";
+    private static final String REQUEST_ID_PATTERN = "[^\\n]*";
+    private static final String RESOURCE_PATTERN = "- " + URI_PATTERN;
+    private static final String RESOURCES_PATTERN = "(?:\n" + RESOURCE_PATTERN + ")*";
+    private static final Pattern SIGN_IN_WITH_ETHEREUM_PATTERN = Pattern.compile(
+        "^(?:(?<scheme>" + SCHEME_PATTERN + ")://)?(?<domain>" + DOMAIN_PATTERN + ") wants you to sign in with your Ethereum account:\n" +
+            "(?<address>" + ADDRESS_PATTERN + ")\n\n" +
+            "(?<statement>" + STATEMENT_PATTERN + ")?\n\n" +
+            "URI: (?<uri>" + URI_PATTERN + ")\n" +
+            "Version: (?<version>" + VERSION_PATTERN + ")\n" +
+            "Chain ID: (?<chainId>" + CHAIN_ID_PATTERN + ")\n" +
+            "Nonce: (?<nonce>" + NONCE_PATTERN + ")\n" +
+            "Issued At: (?<issuedAt>" + DATE_TIME_PATTERN + ")" +
```

Comment:
Are you sure these fields must be exactly in this order?

A message with `URI` and `Version` swapped seems valid, but would not match this regexp.

Reply:
They must be in this order, there is an ABNF here https://eips.ethereum.org/EIPS/eip-4361 which specifies this.